services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpReviewProfileAuthenticator.java

```diff
@@ -110,28 +127,37 @@ protected void actionImpl(AuthenticationFlowContext context, SerializedBrokeredI
             return;
         }
 
-        String username = realm.isRegistrationEmailAsUsername() ? formData.getFirst(UserModel.EMAIL) : formData.getFirst(UserModel.USERNAME);
-        userCtx.setUsername(username);
-        userCtx.setFirstName(formData.getFirst(UserModel.FIRST_NAME));
-        userCtx.setLastName(formData.getFirst(UserModel.LAST_NAME));
+        UserProfileUpdateHelper.update(UserUpdateEvent.IdpReview, context.getSession(), new UserModelDelegate(null) {
```

Comment:
I think `UserProfileUpdateHelper` can be simpler if we just pass the attribute map instead of the `UserModel`. It would avoid having to create this anonymous class to work around the contract to only fetch attributes.

I'm also wondering why the logic on `UserProfileUpdateHelper` is not on the `UserProfileProvider` itself?

Reply:
Regards the position of user profile itself see other comment.

Regards the AttributeList.
This won't work because of the nature of the different UserModel implementations (e,g, for idpReview). Some just are not useable as simple attribute list. (Their interface for the attributes does not work consistently)

Reply:
The differences are reflected in the different user profile implementations